ibis/backends/__init__.py

```diff
@@ -473,6 +473,46 @@ def to_parquet(
                 for batch in batch_reader:
                     writer.write_batch(batch)
 
+    @util.experimental
+    def to_parquet_dir(
+        self,
+        expr: ir.Table,
+        path: str | Path,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Write the results of executing the given expression to a parquet file in a directory.
+
+        This method is eager and will execute the associated expression
+        immediately.
+
+        Parameters
+        ----------
+        expr
+            The ibis expression to execute and persist to parquet.
+        path
+            The data source. A string or Path to the directory where the parquet file will be written.
+        params
+            Mapping of scalar parameter expressions to value.
+        **kwargs
+            Additional keyword arguments passed to pyarrow.parquet.ParquetWriter
+
+        https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html
+
+        """
+        self._import_pyarrow()
+        import pyarrow.parquet as pq
+
+        dir_path = Path(path)
+        dir_path.mkdir(parents=True, exist_ok=True)
+
+        with expr.to_pyarrow_batches(params=params) as batch_reader:
+            for i, batch in enumerate(batch_reader):
+                file_path = f"{dir_path}/out_{i}.parquet"
+                with pq.ParquetWriter(file_path, batch.schema, **kwargs) as writer:
+                    writer.write_batch(batch)
```

Comment:
This whole thing can be reduced quite a bit by using pyarrow datasets.

You should be able to also remove to `mkdir` call above, since `write_dataset` appears to already do that.

```suggestion
        import pyarrow.dataset as ds
        with expr.to_pyarrow_batches(params=params) as batch_reader:
            ds.write_dataset(batch_reader, base_dir=dir_path, format="parquet", **kwargs)
```